### PR TITLE
Multiple nodal d.o.f.

### PR DESCRIPTION
**Pull request by chleh**
This PR is a reincarnation of #821.

It allows for using multiple degrees of freedom per node.

Changes are:
* VectorMatrixAssembler computes local nodal values for each degree of freedom and passes them on to the local assembler.
* Thus, local assemblers have to accept parameters: `assemble(x, xold)`
* MeshComponentMap stores number of components and global matrix order (by location vs. by component)
* some methods now have component id and dof_table as additional input parameters.
* dof table now saves global indices in a matrix indexed by mesh element and component id

TODOs (will not be handled in this PR):
* VectorMatrixAssembler passes all local nodal values in a single vector. In the future that might be changed to passing each component separately.
* using `LocalToGlobalIndexMap::deriveBoundaryConstrainedMap()` on a derived dof table is allowed only if certain conditions hold (see doc in source file). Yet, that might be a rather rare case to use.

### CODE REVIEW

File: AssemblerLib/MeshComponentMap.cpp [hunk 1]
```diff
@@ -259,4 +263,34 @@ std::vector<GlobalIndexType> MeshComponentMap::getGlobalIndicesByComponent(
     return global_indices;
 }
 
+std::vector<GlobalIndexType>
+MeshComponentMap::getIndicesForComponent(
+        const std::vector<GlobalIndexType>& cnt,
+        const unsigned component_id) const
+{
+    std::vector<GlobalIndexType> new_cnt;
+    auto const N = cnt.size() / _num_components;
```
Comment (wenqing) [inline AssemblerLib/MeshComponentMap.cpp]: In some case, DOF per node can vary. For example, using linear finite element for one PDE (e.g. that of fluid flow) and high order finite element for another PDE (e.g. that of solid deformation) in one geometry element is needed for solving of some coupled equations.

File: ProcessLib/ProcessVariable.h [hunk 2]
```diff
@@ -60,8 +60,9 @@ class ProcessVariable
 	MeshLib::Mesh const& getMesh() const;
 
 	void initializeDirichletBCs(MeshGeoToolsLib::MeshNodeSearcher& searcher,
-	                            std::vector<std::size_t>& global_ids,
-	                            std::vector<double>& values);
+	        const AssemblerLib::LocalToGlobalIndexMap& local2global,
+	        const unsigned nodal_dof_idx,
+	        std::vector<std::size_t>& global_ids, std::vector<double>& values);
```
Comment (wenqing) [inline ProcessLib/ProcessVariable.h]: When you merge or rebase, you need to change ``td::size_t`` here to ``GlobalIndexType``.

File: AssemblerLib/VectorMatrixAssembler.h [hunk 3]
```diff
@@ -49,15 +56,46 @@ class VectorMatrixAssembler
...
+        auto const& x = *_x;
+        auto const& x_pts = *_x_prev_ts;
+
+        const unsigned num_comp = _data_pos.getNumComponents();
+
+        if (_x)         localX.reserve(element_dof);
+        if (_x_prev_ts) localX_pts.reserve(element_dof);
+        remapped_rows_cols.reserve(element_dof);
+
+        // The local matrix will always be ordered by component,
```
Comment (chleh) [inline AssemblerLib/VectorMatrixAssembler.h]: Yes, vector might be more appropriate, here. But also the local matrix has to be ordered by component due to the `addToGlobal()` call below.

File: AssemblerLib/MeshComponentMap.cpp (hunk 1 above)
Comment (chleh) [inline AssemblerLib/MeshComponentMap.cpp]: That's a valid point. So far I've only considered scalar components which all operate on basically the same mesh, i.e. all have same order finite elements.
I will think about that.

**Comment by chleh** [comment]
Test has been updated: Less indentation, more independent cases.

File: Tests/AssemblerLib/LocalToGlobalIndexMapMultiComponent.cpp [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+
+	std::size_t operator()(std::size_t const node,
+	                       std::size_t const component) const
+	{
+		return node * num_components + component;
+	}
+};
+
+
+template <enum AL::ComponentOrder ComponentOrder>
```
Comment (endJunction) [inline Tests/AssemblerLib/LocalToGlobalIndexMapMultiComponent.cpp]: My fault here:
```
F:\jenkins\workspace\OGS-6\Win-PRs\ogs\Tests\AssemblerLib\LocalToGlobalIndexMapMultiComponent.cpp(166): error C3431: 'ComponentOrder' : a scoped enumeration cannot be redeclared as an unscoped enumeration [F:\jenkins\workspace\OGS-6\Win-PRs\build\Tests\testrunner.vcxproj]
```

@chleh could you remove the `enum` in templat argument list please.

File: AssemblerLib/LocalToGlobalIndexMap.h [hunk 5]
```diff
@@ -86,27 +95,31 @@ class LocalToGlobalIndexMap
...
     /// Vector contains for each element a vector of global row/or entry indices
     /// in the global stiffness matrix or vector
-    std::vector<LineIndex> _rows;
+    /// first index: element, second index: component
+    Table _rows;
 
     /// Vector alias to that contains for each element a vector of global column
     /// indices in the global stiffness matrix
-    std::vector<LineIndex> const& _columns = _rows;
+    Table const& _columns = _rows;
```
Comment (endJunction) [inline AssemblerLib/LocalToGlobalIndexMap.h]: The comments (the previously bad spelling aside) needs an update.

**Comment by endJunction** [comment]
When building with PETSc following errors emerge:
```
/home/naumov/w/ogs/AssemblerLib/VectorMatrixAssembler.h:79:49: error: type 'const MathLib::PETScVector' does not provide a subscript operator
            if (_x)         localX.emplace_back((*_x)[i]);
                                                ^~~~~ ~
/home/naumov/w/ogs/AssemblerLib/SerialExecutor.h:40:13: note: in instantiation of function template specialization 'AssemblerLib::VectorMatrixAssembler<MathLib::PETScMatrix,
      MathLib::PETScVector>::operator()<ProcessLib::GroundwaterFlow::LocalAssemblerDataInterface<MathLib::PETScMatrix, MathLib::PETScVector> >' requested here
            f(i, c[i], std::forward<Args_>(args)...);
            ^
/home/naumov/w/ogs/AssemblerLib/GlobalSetup.h:46:26: note: in instantiation of function template specialization
      'AssemblerLib::SerialExecutor::execute<AssemblerLib::VectorMatrixAssembler<MathLib::PETScMatrix, MathLib::PETScVector>,
      std::__debug::vector<ProcessLib::GroundwaterFlow::LocalAssemblerDataInterface<MathLib::PETScMatrix, MathLib::PETScVector> *,
      std::allocator<ProcessLib::GroundwaterFlow::LocalAssemblerDataInterface<MathLib::PETScMatrix, MathLib::PETScVector> *> >>' requested here
        return Executor::execute(std::forward<Args>(args)...);
                         ^
/home/naumov/w/ogs/ProcessLib/GroundwaterFlowProcess.h:278:23: note: in instantiation of function template specialization
      'AssemblerLib::GlobalSetup<AssemblerLib::VectorMatrixBuilder<MathLib::PETScMatrix, MathLib::PETScVector>, AssemblerLib::SerialExecutor,
      MathLib::PETScLinearSolver>::execute<AssemblerLib::VectorMatrixAssembler<MathLib::PETScMatrix, MathLib::PETScVector> &,
      std::__debug::vector<ProcessLib::GroundwaterFlow::LocalAssemblerDataInterface<MathLib::PETScMatrix, MathLib::PETScVector> *,
      std::allocator<ProcessLib::GroundwaterFlow::LocalAssemblerDataInterface<MathLib::PETScMatrix, MathLib::PETScVector> *> > &>' requested here
        _global_setup.execute(*_global_assembler, _local_assemblers);
                      ^
In file included from /home/naumov/w/ogs/ProcessLib/GroundwaterFlowProcess.cpp:10:
In file included from /home/naumov/w/ogs/ProcessLib/GroundwaterFlowProcess-fwd.h:13:
In file included from /home/naumov/w/ogs/ProcessLib/GroundwaterFlowProcess.h:28:
/home/naumov/w/ogs/AssemblerLib/VectorMatrixAssembler.h:80:53: error: type 'const MathLib::PETScVector' does not provide a subscript operator
            if (_x_prev_ts) localX_pts.emplace_back((*_x_prev_ts)[i]);
                                                    ^~~~~~~~~~~~~ ~
2 errors generated.
```

File: AssemblerLib/MeshComponentMap.h [hunk 6]
```diff
@@ -46,7 +46,18 @@ class MeshComponentMap
         ComponentOrder order);
 
     /// Creates a subset of the current mesh component map.
-    /// The order of components is the same as of the current map.
+    /// The components are those from \c components which are not nullptrs.
+    /// The order (BY_LOCATION/BY_NODE) of components is the same as of the current map.
```
Comment (endJunction) [inline AssemblerLib/MeshComponentMap.h]: `BY_NODE` is confusing because such order does not exist.
Would simply keep `BY_LOCATION`.

File: ProcessLib/UniformDirichletBoundaryCondition.h [hunk 7]
```diff
@@ -48,13 +49,27 @@ class UniformDirichletBoundaryCondition
...
+            std::vector<double>& values)
     {
         // Find nodes' ids on the given mesh on which this boundary condition
         // is defined.
         std::vector<std::size_t> ids = searcher.getMeshNodeIDs(*_geometry);
 
+        // convert mesh node ids to global index for the given component
+        for (auto& id : ids)
+        {
+            MeshLib::Location l(searcher.getMesh().getID(),
```
Comment (endJunction) [inline ProcessLib/UniformDirichletBoundaryCondition.h]: I would replace the `searcher.getMesh().getID()` by `searcher.getMeshId()` just forwarding that id and not exposing the mesh of the searcher. If I didn't miss something, this is the only place where node searcher's `getMesh()` is used.

File: AssemblerLib/VectorMatrixAssembler.h [hunk 8]
```diff
@@ -49,15 +57,41 @@ class VectorMatrixAssembler
...
+
+        std::vector<double> localX;
+        std::vector<double> localX_pts;
+
+        if (_x)         localX.reserve(indices.size());
+        if (_x_prev_ts) localX_pts.reserve(indices.size());
+
+        for (auto i : indices)
+        {
+            if (_x)         localX.emplace_back(_x->get(i));
```
Comment (endJunction) [inline AssemblerLib/VectorMatrixAssembler.h]: Note for later:
The get() call for negative indices in case of PETSc vectors will not return the expected value in all cases.
See [PETSc::VecGetValues](http://www.mcs.anl.gov/petsc/petsc-current/docs/manualpages/Vec/VecGetValues.html) documentation.

File: AssemblerLib/VectorMatrixAssembler.h [hunk 9]
```diff
@@ -38,6 +38,14 @@ class VectorMatrixAssembler
 
     ~VectorMatrixAssembler() {}
 
+    void setX(GLOBAL_VECTOR_ const * x, GLOBAL_VECTOR_ const * x_prev_ts)
+    {
+        assert((!x == !x_prev_ts) && "either no or both inputs have to be set");
+        assert((!x) || x->size() == x_prev_ts->size());
+        _x = x;
+        _x_prev_ts = x_prev_ts;
+    }
```
Comment (endJunction) [inline AssemblerLib/VectorMatrixAssembler.h]: Passing the solution `x` to assembler is required for the non-linear assembly; OK.
How is the `x_prev_ts` used?

Comment (chleh) [inline AssemblerLib/VectorMatrixAssembler.h]: I use `x_prev_ts` for the assembly of the global equation system. E.g. for a parabolic problem the linear equation system to solve is `M x' + K x = b`, where `x' = (x - y) / dt`
So (currently) in need both `x` and `y := x_prev_ts`.
Maybe later on this can be moved to a different place.

**Comment by endJunction** [comment]
:fast_forward:

File: AssemblerLib/VectorMatrixAssembler.h (hunk 8 above)
Comment (wenqing) [inline AssemblerLib/VectorMatrixAssembler.h]: Get values entry by entry is slow in PETSc. There is a member in class PETScVector called  
```
get(const std::vector<PetscInt> &e_idxs,  T_SUBVEC &sub_vec)
```
You can try to perform it like
```
_x->get(indices, localX);
```
For this purpose, you need to add        
```
 template<class T_SUBVEC> void get(const std::vector<[Int Type of the owner package]> &e_idxs,
                                          T_SUBVEC &sub_vec)
```
to all vector classes in MathLib/LinAlg. 

Besides, do not forget to set  indices be positive.

File: AssemblerLib/VectorMatrixAssembler.h [hunk 10]
```diff
@@ -49,15 +57,41 @@ class VectorMatrixAssembler
...
 
-        LocalToGlobalIndexMap::RowColumnIndices const& indices = _data_pos[id];
+        std::vector<GlobalIndexType> indices;
+
+        // Local matrices and vectors will always be ordered by component,
+        // no matter what the order of the global matrix is.
+        for (unsigned c=0; c<_data_pos.getNumComponents(); ++c)
+        {
+            auto const& idcs = _data_pos(id, c).rows;
+            indices.reserve(indices.size() + idcs.size());
```
Comment (wenqing) [inline AssemblerLib/VectorMatrixAssembler.h]: set fabs indices

File: AssemblerLib/LocalToGlobalIndexMap.h [hunk 11]
```diff
@@ -86,27 +95,30 @@ class LocalToGlobalIndexMap
...
 
 private:
-    std::vector<MeshLib::MeshSubsets*> const& _mesh_subsets;
+    std::vector<MeshLib::MeshSubsets*> const _mesh_subsets;
     AssemblerLib::MeshComponentMap _mesh_component_map;
 
-    /// Vector contains for each element a vector of global row/or entry indices
-    /// in the global stiffness matrix or vector
-    std::vector<LineIndex> _rows;
+    using Table = Eigen::Matrix<LineIndex, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>;
```
Comment (wenqing) [inline AssemblerLib/LocalToGlobalIndexMap.h]: It is possible to place this alias some where outside of this file, e.g. a a file like ``MathLib/LinAlg/MatrixTools.h`` for comment access? Under such consideration, we can replace the eigen with other third party algebraic package easily.

Comment (chleh) [inline AssemblerLib/LocalToGlobalIndexMap.h]: I think, currently there are no benefits from putting this definition to some other place: It is an implementation detail of the dof table and only serves as a special lookup table having two indices.
But maybe I should define it in terms of either `Eigen::Matrix` or `DenseMatrix` depending on the setting `OGS_USE_EIGEN`.

File: AssemblerLib/VectorMatrixAssembler.h (hunk 8 above)
Comment (chleh) [inline AssemblerLib/VectorMatrixAssembler.h]: > Get values entry by entry is slow in PETSc.

Is there really a big performance impact from getting values index-by-index? If I use `get(const std::vector<PetscInt> &e_idxs,  T_SUBVEC &sub_vec)`, the PETSc function probably is faster (though, not with zero runtime), but results have to be cached.

Additionally, extending all vector classes would clutter this PR. Sure, this is a rather weak argument. But, as @endJunction and you pointed out, there is an issue with negative indices and PETSc. This is definitely out of scope of this PR. I think ghost nodes should be handled transparently in the class `PetscVector`.

I suggest to leave this issue open for this PR and to work on a solution for it afterwards.

Do you agree?

File: AssemblerLib/VectorMatrixAssembler.h (hunk 10 above)
Comment (chleh) [inline AssemblerLib/VectorMatrixAssembler.h]: Excuse me, I don't understand what you mean. Could you explain again, please?

Comment (wenqing) [inline AssemblerLib/VectorMatrixAssembler.h]: Just related the negative ID for PETSc, which you mentioned to put it in the next PR in your reply to the next comment.

File: AssemblerLib/VectorMatrixAssembler.h (hunk 8 above)
Comment (wenqing) [inline AssemblerLib/VectorMatrixAssembler.h]: In PETScVector, get(PestcInt idx) calls PETSc routine
```
VecGetValues(_v, 1, &idx, &x);
```
e.g. it get one entry through an array with the size of one.
To fetch values from petsc vector, communication operation is involved among node. To get a bunch of values through an array in one time costs much less time than to get it one by one.

Comment (chleh) [inline AssemblerLib/VectorMatrixAssembler.h]: OK, but I think extending (and unifying) matrix classes' interfaces should also be done in a separate PR.

Comment (wenqing) [inline AssemblerLib/VectorMatrixAssembler.h]: That's good.

**Comment by wenqing** [comment]
:+1:

File: Tests/AssemblerLib/LocalToGlobalIndexMapMultiComponent.cpp [hunk 12]
```diff
@@ -0,0 +1,277 @@
...
+	}
+
+	template <AL::ComponentOrder order>
+	void test(const unsigned num_components, const unsigned selected_component,
+	          std::function<std::size_t(std::size_t, std::size_t)> const
+	              compute_global_index);
+
+	std::unique_ptr<const MeshLib::Mesh> mesh;
+	std::unique_ptr<const MeL::MeshSubset> mesh_items_all_nodes;
+    std::vector<MeL::MeshSubsets*> components_boundary;
```
Comment (TomFischer) [inline Tests/AssemblerLib/LocalToGlobalIndexMapMultiComponent.cpp]: Can you check the indentation?

**Comment by TomFischer** [comment]
:+1:
